pkg/checks/traceroute/traceroute.go

```diff
@@ -2,197 +2,327 @@ package traceroute
-// This is also called while the check is running, if the remote config is updated
-// This should return an error if the config is invalid
-func (tr *Traceroute) SetConfig(cfg checks.Runtime) error {
-	if cfg, ok := cfg.(*Config); ok {
-		tr.Mu.Lock()
-		defer tr.Mu.Unlock()
-		tr.config = *cfg
-		return nil
+func handleTcpSuccess(conn net.Conn, addr net.Addr, ttl int, latency time.Duration) *Hop {
+	conn.Close() // #nosec G104
+
+	ipaddr := ipFromAddr(addr)
+	names, _ := net.LookupAddr(ipaddr.String()) // we don't care about this lookup failing
+
+	name := ""
+	if len(names) >= 1 {
+		name = names[0]
+	}
+
+	return &Hop{
+		Latency: latency,
+		Ttl:     ttl,
+		Addr:    newHopAddress(addr),
+		Name:    name,
+		Reached: true,
+	}
+}
+
+// handleIcmpResponse attempts to read a time exceeded packet that matches clientPort until timeout is reached
+// if an error occurs while reading from the socket, handleIcmpResponse will silently fail and return a hop with hop.Reached=false
+func handleIcmpResponse(ctx context.Context, icmpListener *icmp.PacketConn, clientPort, ttl int, timeout time.Duration) Hop {
+	log := logger.FromContext(ctx)
+	deadline := time.Now().Add(timeout)
+
+	for time.Now().Unix() < deadline.Unix() {
+		log.Debug("Reading ICMP message", "ttl", ttl)
+		gotPort, addr, err := readIcmpMessage(ctx, icmpListener, timeout)
+		if err != nil {
+			log.Debug("Failed to read ICMP message", "err", err.Error(), "ttl", ttl)
+			continue
+		}
+
+		// Check if the destination port matches our dialer's source port
+		if gotPort == clientPort {
+			ipaddr := ipFromAddr(addr)
+			names, _ := net.LookupAddr(ipaddr.String()) // we don't really care if this lookup works, so ignore the error
+
+			name := ""
+			if len(names) >= 1 {
+				name = names[0]
+			}
+
+			return Hop{
+				Ttl:  ttl,
+				Addr: newHopAddress(addr),
+				Name: name,
+			}
+		}
 	}
 
-	return checks.ErrConfigMismatch{
-		Expected: CheckName,
-		Current:  cfg.For(),
+	log.Debug("Deadline reached", "ttl", ttl)
+	return Hop{
+		Ttl: ttl,
 	}
 }
 
-// Schema returns an openapi3.SchemaRef of the result type returned by the check
-func (tr *Traceroute) Schema() (*openapi3.SchemaRef, error) {
-	return checks.OpenapiFromPerfData[map[string]result](map[string]result{})
+type Hop struct {
+	Latency time.Duration `json:"latency" yaml:"latency" mapstructure:"latency"`
+	Addr    HopAddress    `json:"addr" yaml:"addr" mapstructure:"addr"`
+	Name    string        `json:"name" yaml:"name" mapstructure:"name"`
+	Ttl     int           `json:"ttl" yaml:"ttl" mapstructure:"ttl"`
+	Reached bool          `json:"reached" yaml:"reached" mapstructure:"reached"`
 }
 
-// GetMetricCollectors allows the check to provide prometheus metric collectors
-func (tr *Traceroute) GetMetricCollectors() []prometheus.Collector {
-	return []prometheus.Collector{}
+type HopAddress struct {
+	IP   string `json:"ip" yaml:"ip" mapstructure:"ip"`
+	Port int    `json:"port" yaml:"port" mapstructure:"port"`
```

Comment:
I think it's appropriate to use a `uint16` type here

Reply:
I didn't use a uint16, because the [standard](https://cs.opensource.google/go/go/+/go1.22.5:src/net/tcpsock.go;l=21) [library](https://cs.opensource.google/go/go/+/refs/tags/go1.22.5:src/net/udpsock.go;l=24) doesn't either:

```go
type UDPAddr struct {
	IP   IP
	Port int
	Zone string // IPv6 scoped addressing zone
}

// TCPAddr represents the address of a TCP end point.
type TCPAddr struct {
	IP   IP
	Port int
	Zone string // IPv6 scoped addressing zone
}
```